clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -1555,11 +1560,14 @@ scheduler_tmpl_t<RecordType, ReaderType>::pick_next_input(output_ordinal_t outpu
                     else
                         index = prev_index; // Go back to prior.
                 } else {
-                    // Give up the input before we go to the queue so we can add
-                    // ourselves to the queue.  If we're the highest priority we
-                    // shouldn't switch.  The queue preserves FIFO for same-priority
-                    // cases so we will switch if someone of equal priority is waiting.
-                    set_cur_input(output, INVALID_INPUT_ORDINAL);
+                    if (!in_wait_state) {
+                        // Give up the input before we go to the queue so we can add
+                        // ourselves to the queue.  If we're the highest priority we
+                        // shouldn't switch.  The queue preserves FIFO for same-priority
+                        // cases so we will switch if someone of equal priority is
+                        // waiting.
+                        set_cur_input(output, INVALID_INPUT_ORDINAL);
```

Comment:
What if the only thread that's left is the thread that executed the blocking system call?

Reply:
In that case it will go into the `if (ready_queue_empty())` condition right above this one.  I went ahead and added a test of just one thread with blocking syscalls just to confirm there's no hang or anything.